### DISCUSSION TITLE
Analyze PDFs and Documents

### DISCUSSION THREAD

**Discussion by DanielusG** [oobabooga/text-generation-webui#1372] [General]
it would be great if there was an extension capable of loading documents, and with the long term memory extension remember it and be able to ask questions about it

There is a way to do it?

**Comment by Felixqian4160** [comment]
I want to know too

**Reply by TomLucidor** [reply]
Currently digging into this instead https://github.com/oobabooga/text-generation-webui/discussions/1933

**Comment by syddharth** [comment]
I did this locally with a notebook using Langchain and LlamaIndex. Maybe someone could help make this into an extention.

Here is the code I used (I have very basic novice knowledge in Python):

```
# Load Libraries
import torch
from transformers import LlamaTokenizer, LlamaForCausalLM, GenerationConfig, pipeline

from langchain.llms import HuggingFacePipeline
from langchain import PromptTemplate, LLMChain
from langchain.llms.base import LLM
from langchain.embeddings.huggingface import HuggingFaceEmbeddings

from llama_index import LangchainEmbedding
from llama_index import SimpleDirectoryReader, LLMPredictor, GPTSimpleVectorIndex, ServiceContext, PromptHelper
from llama_index import download_loader

from pathlib import Path

# Load Model
vicuna_PATH = r'C:\AI\T2T\_source\models\vicuna-7b'
model_id = vicuna_PATH

tokenizer = LlamaTokenizer.from_pretrained(model_id)
model = LlamaForCausalLM.from_pretrained(
    model_id,
    device_map='auto',
    load_in_8bit=True,
    torch_dtype=torch.float16
)

# Create Class for llama_index
class LLaMALLM(LLM):
    def _call(self, prompt, stop=None):
        prompt += "\n### Assistant:Sure,"
 
        inputs = tokenizer(prompt, return_tensors="pt")
        input_ids = inputs["input_ids"].cuda()

        generation_config = GenerationConfig(
                max_length=128,
                do_sample=True,
                top_p=0.1,
                top_k=40,
                temperature=0.7,
                repetition_penalty=1.18,
                typical_p=1.0
         )
        with torch.no_grad():
            generation_output = model.generate(
                input_ids=input_ids,
                generation_config=generation_config,
                return_dict_in_generate=True,
                output_scores=True,
                max_new_tokens=128,
            )
        response = ""
        for s in generation_output.sequences:
            response += tokenizer.decode(s)
            
        response = response[len(prompt):]
        print("Model Response:", response)
        return response
 
    def _identifying_params(self):
        return {"name_of_model": "alpaca"}
 
    def _llm_type(self):
        return "custom"

llm_predictor = LLMPredictor(llm=LLaMALLM())

embed_model = LangchainEmbedding(HuggingFaceEmbeddings())

# Convert Document into Index
EpubReader = download_loader("EpubReader")
loader = EpubReader()

documents = loader.load_data(file=Path('./data/book.epub'))
documents = SimpleDirectoryReader('data').load_data()


service_context = ServiceContext.from_defaults(llm_predictor=llm_predictor, embed_model=embed_model)

index = GPTSimpleVectorIndex.from_documents(documents, service_context=service_context)
index.save_to_disk('./index/index_epub01.json')


response = index.query("What is the document about?")
print(response.response)

# Load from Existing Index
new_service_context = ServiceContext.from_defaults(llm_predictor=llm_predictor, embed_model=embed_model)
new_index = GPTSimpleVectorIndex.load_from_disk('./index/index_epub02.json', service_context=new_service_context)

# Ask LLM about document
response = new_index.query("What is the best way to live?")
print(response.response)
```

One would need to change the vicuna_PATH and the path to epub.

**Comment by BarfingLemurs** [comment]
I think you just have to enter all the documents repeatedly as input in the long term memory extension, or find a script that can convert those documents to the text file format used by the long term memory extension

**Reply by DanielusG** [reply]
I have also thought about this hypothesis (I also opened an issue in the long_term_memory extension)

I tried pasting the whole document into the AI chat, but (probably due to my computer not having enough power) it goes into CUDA Out of memory...

Even if I have 6GB of VRAM and I use a small quantized model, like vicuna 7B which uses only 3GB of VRAM, still pasting the whole document goes into out of memory.

So I would need a script that directly puts the document into the long_term_memory, without going through the chat

**Reply by BarfingLemurs** [reply]
I tried the extension with gpt-x-alpaca just now. I couldn't make it repeat verbatium what was in memory, but it's useful as the embedding will be auto-loaded by the extension if you are discussing a relevant topic. If I discuss tacos, it won't load it in.
![image](https://user-images.githubusercontent.com/128182951/234252833-d9c33fc6-7b9a-40e4-a2ae-5864768898e1.png)


The extension stores any especially long user input in a database file (.db) viewable here: https://sqliteviewer.app
it doesn't seem to take extra vram, just cpu ram (nice)
The extension seems to be using a model downloaded from huggingface during install, which creates embeddings of the raw text, seen in the sql database.
The embedding(s) are /home/ubuntu/oobabooga_linux/text-generation-webui/extensions/long_term_memory/user_data/bot_memories/long_term_memory_embeddings.zarr/0.0

**Reply by DanielusG** [reply]
I noticed that as you say long_term_memory does not actually use VRAM. What uses all my VRAM is the model that tries to interpret what I just posted in the chat
So you need a way to have only long_term_memory save the data, without then having the model respond.
Although I think that when you then ask a question about that topic, long_term_memory will load the entire document in context and it will go out of memory (I'm talking about my problem with vram, not in general)

**Reply by Felixqian4160** [reply]
If you first use Langchain to save the content of the document, then generate a small LoRA, and every time you read the contents of the document, put it in this LoRA model. Will it reduce the usage of storage?

**Reply by DanielusG** [reply]
But for train a lora you need a lot of power and a good GPU...

**Reply by BarfingLemurs** [reply]
you will get the speakers ideas and combined with the context (which you can see the exact saved text brought up as a setting)
There is a script in the issues page that can save all your new lines as embeddings and to database.

**Reply by BarfingLemurs** [reply]
lora in 4bit maybe you can use a colab instance and a smaller model, I'm not sure if training works with 4-6gb, havent tested all the settings

**Reply by kalle07** [reply]
hello i wait also to add some PDF to a model !
why can GPT4ALL can do this ... but its on use mainly CPU ? abd its 10 times slower than oobd   ... ?

**Reply by creeduk** [reply]
This is also possible with PrivateGPT I saw this originally in a Aitrepreneur video. This was the git page  -  https://github.com/imartinez/privateGPT. It would be good as this is to be the A1111 of text gen then an extension or way to have source docs and "talk to them" would be awesome.
